Issue #35 / flipNormal()

otmanesabir opened the pull request:
As mentioned in the previous PR, we'll be splitting this task into two. This is the first part.


Partially Resolves #35

otmanesabir committed:
Issue #35 / flipNormal() Part 1

ereator commented on modules/core/IPrim.h:
```diff
@@ -99,10 +81,41 @@ namespace rt {
-		std::string			m_name;				///< Optional name of the primitive
+		const ptr_shader_t	m_pShader;			///< Pointer to the shader, see @ref  IShader.
+		std::string			m_name;				///< Optional name of the primitive.
+		bool			    m_flipped;			///< Flag which helps decide whether to flip the normal or not.
```
We should not forget to initialize it to _false_ by defaut
`bool m_flipped = false;`

ereator reviewed (changes requested):
Thank you so much for nice and clear implementation.
I have started checking yesterday the new functionality in this PR and found two problems, that we did not foreseen:
1. The inverting normals of a Solid does not make any effect. This is possible to fix in a very straightforward way, so I would not say it is a _big_ problem
2. The _larger_ problem is that inversion of the normals of a composite does not work as well and I do not know at the moment how to fix it. As the method `CCompositeGeometry::doGetNormal()` should not be ever called, flipping normals of a composite will never take effect. The brute-force idea to fix that is to make method `virtual void IPrim::flipNormals() = 0;` to be pure virtual and force all derived classes to implement it. Any better idea is highly appreciated...

ereator committed:
add flipNormal() to the Solid

ereator committed:
Update IPrim.h

1. Setting the m_flipped to be false on initialization
2. Making the flipNormal() method to be virtual (so in composite it could be overidden)

ereator committed:
Add flipNormals() to the Composite

ereator commented:
I have fixed the first issue, I mentioned above and made the flipped flag to be false on initialisation. I have also added the implementation of the `flipNormal()` method to the CPrimComposite. The preliminary implementation is just the same as for the solids - flipping the normals of all the primitives in `m_vPrims1` and `m_vPrims2`. However, the result of _e.g._ substraction becomes incorrect. The logic of entering and exiting the primitives is affected, thus flipping normals on composite does not only flips the normals, but also changes the entire result. We need to think how to fix that.

otmanesabir commented:
thank you for fixing the first issue!
However, I still believe flipping the composite probably won't fix this issue.
I have an idea on how to potentially fix this issue but first I can try to explain what the problem is.

When we flip a primitive in the CSG tree, we do it 'globally'. Therefore, on the next ray intersection, that composite is still flipped. This is not too bad; however, it causes a problem on the next ray evaluation. Basically, when we want to evaluate if a ray is "entering" or "exiting" the primitive, we rely on the normal to figure that out. However, since we flip the normal on the previous ray, then this evaluation is done wrongly because now the `getNormal(..)` method returns the inverted normal. This is not good. Because all subsequent evaluations are also messed up. I think we can have two solutions:

1. Call `doGetNormal()` in `CCompositeGeometry` ray evaluation to get the **original** normal instead this flipped one 
2. Create a getter `isFlipped()` and factor the result of this boolean when we're doing the ray evaluation. 

I believe this should fix this issue but could still cause artifacts when dealing with larger primitives:

Let's imagine a case where a prim sphere is doing union, intersection, and difference. The difference operation is first and flips the normal of this entire sphere and now the union and intersection operations will run into consideration that this entire solid is pointing differently to where it's supposed to be. Therefore, generally speaking, I'm not sure if adding this on a primitive level is the solution for removing the Dummy primitives.

ereator committed:
Fix: flipping normals in Composite

ereator commented:
I have tried out the idea number 2 from your comment and it gave a correct results on the earth scene! 
I have implemented it in a fast manner, so without getter `isFlipped()` but by introducing another `bool m_isFlipped = false` in the composite class. 
I will try to perform additional tests and if everything now running correctly, I believe we can clean-up the code and merge the PR :-)

otmanesabir commented:
very nice! I'm very glad it worked! 
I would be happy to also further test this over the coming weekend and if everything works well, can also clean up and attach a full description with my test scenes :D

otmanesabir committed:
Issue #35 Remove PrimDummy

otmanesabir commented:
Implemented this on the primitive level and cleaned up the implementation a bit. I also got rid of this unused PrimDummy now :)


Attached is a scene I tested with also multiple operations. (Primitives and Solids). I think all looks good now but I could've missed something so would also wait for your confirmation. We can also think of adding unit tests for this composite logic to make sure that no changes break it. :) 


<img width="795" alt="Screenshot 2021-12-04 at 12 07 48" src="https://user-images.githubusercontent.com/41652846/144724044-846fd4c1-c42c-4caf-b84a-913ac0d5b94e.png">

ereator commented on modules/core/Ray.h:
```diff
@@ -18,7 +18,7 @@ namespace rt {
 		size_t 							counter;											///< Number of re-traces
 		
 		double							t		= std::numeric_limits<double>::infinity();	///< Current/maximum hit distance
-		std::shared_ptr<const IPrim>	hit		= nullptr;									///< Pointer to currently closest primitive
```
Is it possible to leave it const ?

ereator commented on modules/core/PrimTriangle.h:
```diff
@@ -44,12 +44,12 @@ namespace rt {
 		{}
 		DllExport virtual ~CPrimTriangle(void) = default;
 		
-		DllExport virtual bool	intersect(Ray& ray) const override;
```
is it possible to leave it as const ?

ereator commented on modules/core/PrimTriangle.cpp:
```diff
@@ -3,7 +3,7 @@
 #include "Transform.h"
 
 namespace rt {
-	bool CPrimTriangle::intersect(Ray& ray) const
```
is it possible to leave it as const ?

ereator commented on modules/core/PrimSphere.h:
```diff
@@ -27,11 +27,11 @@ namespace rt {
 		{}
 		DllExport virtual ~CPrimSphere(void) = default;
 
-		DllExport virtual bool 			intersect(Ray& ray) const override;
```
is it possible to leave it as const ?

ereator commented on modules/core/PrimSphere.cpp:
```diff
@@ -4,7 +4,7 @@
 #include "macroses.h"
 
 namespace rt {
-	bool CPrimSphere::intersect(Ray& ray) const
```
is it possible to leave it as const ?

ereator commented on modules/core/PrimSphere.cpp:
```diff
@@ -56,14 +56,14 @@ namespace rt {
 	{
-		return intersect(lvalue_cast(Ray(ray)));
-	}
+        return const_cast<CPrimSphere*>(this)->intersect(lvalue_cast(Ray(ray)));
```
This casting is not quite clear. if there are cases when `this` is not CPromSphere?

ereator commented on modules/core/PrimPlane.h:
```diff
@@ -29,11 +29,11 @@ namespace rt {
 		}
 		DllExport virtual ~CPrimPlane(void) = default;
 
-		DllExport virtual bool 			intersect(Ray& ray) const override;
```
is it possible to leave it as const ?

ereator commented on modules/core/PrimPlane.cpp:
```diff
@@ -3,7 +3,7 @@
 #include "Transform.h"
 
 namespace rt {
-	bool CPrimPlane::intersect(Ray& ray) const
```
is it possible to leave it as const ?

ereator commented on modules/core/PrimDisc.h:
```diff
@@ -29,11 +29,11 @@ namespace rt {
 		{}
 		DllExport virtual ~CPrimDisc(void) = default;
 		
-		DllExport virtual bool				intersect(Ray& ray) const override;
```
is it possible to leave it as const ?

ereator commented on modules/core/PrimDisc.cpp:
```diff
@@ -3,7 +3,7 @@
 #include "Transform.h"
 
 namespace rt {
-	bool CPrimDisc::intersect(Ray& ray) const
```
is it possible to leave it as const ?

ereator commented on modules/core/CompositeGeometry.cpp:
```diff
@@ -28,7 +27,7 @@ namespace rt {
 #endif
     }
 
-    bool CCompositeGeometry::intersect(Ray &ray) const {
```
is it possible to leave it as const ?

ereator commented:
I would try not to make the const methods and pointer in ray to be non-const. Let us first concentrate on inverting normals and then on removing the dummy primitive

ereator committed:
Reverting the changed files

ereator committed:
Revert making intersect() method to be non-const

ereator committed:
Code clean-up

ereator committed:
Move the new methods doGetNormal() and doGetShadingNormal() to be private

ereator committed:
Add flipping normal functionality to composite prmitive

ereator committed:
Add DllExport to the private member-functions doGetNormal(), doGetShadingNormal()

ereator committed:
Code final cleaning
